src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -518,7 +518,17 @@ private function createEncoder($config, ContainerBuilder $container)
         // Argon2i encoder
         if ('argon2i' === $config['algorithm']) {
             if (!Argon2iPasswordEncoder::isSupported()) {
-                throw new InvalidConfigurationException('Argon2i algorithm is not supported. Please install the libsodium extension or upgrade to PHP 7.2+.');
+                // Debugging on Heroku and derivatives is a lot of work, show specific error to save end users some time
+                if (false !== strpos($_ENV['PATH'], 'heroku')) {
+                    throw new InvalidConfigurationException('Argon2i algorithm is not supported on Heroku, use BCrypt instead.');
+                }
+                if (\PHP_VERSION_ID >= 70200) {
+                    // Argon2 support got added in libsodium 1.0.9
+                    if (version_compare(\SODIUM_LIBRARY_VERSION, '1.0.9', '<')) {
```

Comment:
I feel that this is going to far, we won't be able to cover every single edge case that causes argon2i to be unavailable. 
The previous state of this PR was broad enough to me, i.e. `if php >= 7.2 then compile with the proper flag, install the pecl otherwise. Let's see what others think.

Reply:
@chalasr how about doing feature detection instead: https://github.com/symfony/symfony/pull/26190#discussion_r168623244 ?